src/awkward/_errors.py

```diff
@@ -75,50 +69,33 @@ def __enter__(self):
             self._slate.__dict__["__primary_context__"] = self
 
     def __exit__(self, exception_type, exception_value, traceback):
-        try:
+        if (
+            exception_type is not None
+            and issubclass(exception_type, Exception)
+            and self.primary() is self
+        ):
+            # Step out of the way so that another ErrorContext can become primary.
+            # Is this necessary to do here? (We're about to raise an exception anyway)
+            self._slate.__dict__.clear()
             # Handle caught exception
-            if (
-                exception_type is not None
-                and issubclass(exception_type, Exception)
-                and self.primary() is self
-            ):
-                self.handle_exception(exception_type, exception_value)
-        finally:
+            raise self.decorate_exception(exception_type, exception_value)
+        else:
             # Step out of the way so that another ErrorContext can become primary.
             if self.primary() is self:
                 self._slate.__dict__.clear()
 
-    def handle_exception(self, cls: type[E], exception: E):
-        if sys.version_info >= (3, 11, 0, "final"):
-            self.decorate_exception(cls, exception)
-        else:
-            raise self.decorate_exception(cls, exception)
-
     def decorate_exception(self, cls: type[E], exception: E) -> Exception:
-        if sys.version_info >= (3, 11, 0, "final"):
-            if issubclass(cls, (NotImplementedError, AssertionError)):
-                exception.add_note(
-                    "\n\nSee if this has been reported at https://github.com/scikit-hep/awkward/issues"
-                )
+        def _add_note(exception: E, note: str) -> E:
+            if sys.version_info >= (3, 11, 0, "final"):
+                exception.add_note(note)
```

Comment:
Although the pre-PR version of the code had this explicit version number check, I wonder if the check can be more specific and less dependent on external knowledge (the fact that this is the relevant version number and it includes the word "final") with

```suggestion
            if hasattr(exception, "add_note"):
                exception.add_note(note)
```

That is, it seems like we want to start using this method as soon as it exists... unless it existed for a few versions in a broken state that we don't want to use it.

This is just a thought—not a requirement for this PR.

Reply:
I'd recommend dropping the "final" (which doesn't work anyway), but this is better because users can see _why_ `add_note` might be missing, and static tools like Ruff can automatically remove the if statement when the min version is set to 3.11.